src/rust/engine/fs/src/lib.rs

```diff
@@ -455,14 +469,77 @@ impl PosixFS {
       .to_boxed()
   }
 
+  ///
+  /// Makes a Stat for path_for_stat relative to absolute_path_to_root.
+  ///
+  fn stat_internal<F>(
+    path_for_stat: PathBuf,
+    file_type: std::fs::FileType,
+    absolute_path_to_root: &Path,
+    get_metadata: F,
+  ) -> Result<Stat, io::Error>
+  where
+    F: FnOnce() -> Result<fs::Metadata, io::Error>,
+  {
+    if !path_for_stat.is_relative() {
+      return Err(io::Error::new(
+        io::ErrorKind::InvalidInput,
+        format!(
+          "Argument path_for_stat to PosixFS::stat must be relative path, got {:?}",
+          path_for_stat
+        ),
+      ));
+    }
+    if !absolute_path_to_root.is_absolute() {
```

Comment:
Rather than making this method static and then validating that the root is absolute during every stat call, it would probably be good to make this method non-static (ie, borrow self). Since the root has already been canonicalized, you won't need to validate that.

Reply:
I think to do that, I'd need to make `scandir_sync` non-static, and to do that `scandir` would need to be able to move a `&self` into its spawn_fn, and I'm not sure how to do that without Clone-ing PosixFS which would require copying its root, and Arc-wrapping its CpuPool.

Is there a way I can do this without doing that?

Reply:
There is, but let's not worry about it now: basically, by making FS a trait, and then implementing it for Arc<PosixFS>. A TODO is fine.